algorithm/cgal/Pack.h

```diff
@@ -1,49 +1,468 @@
+    const CGAL::Point_2<EK> zero(0, 0);
+
+    sheet_by_input.resize(count);
+
+    std::vector<Sheet> sheets;
+    std::vector<Part> parts;
+
+    orientations.push_back(0);
+
+    auto move_pwhs = [&](std::vector<CGAL::Polygon_with_holes_2<EK>>& pwhs,
+                         const EK::Vector_2& offset) {
+      for (auto& pwh : pwhs) {
+        CGAL::Polygon_2<EK> nb;
+        assert(pwh.outer_boundary().orientation() == CGAL::Sign::POSITIVE);
+        for (auto& point : pwh.outer_boundary()) {
+          nb.push_back(point + offset);
+        }
+        std::vector<CGAL::Polygon_2<EK>> nhs;
+        for (auto& hole : pwh.holes()) {
+          CGAL::Polygon_2<EK> nh;
+          for (auto& point : hole) {
+            assert(hole.orientation() == CGAL::Sign::NEGATIVE);
+            nh.push_back(point + offset);
+          }
+          nhs.push_back(std::move(nh));
+        }
+        pwh = CGAL::Polygon_with_holes_2<EK>(nb, nhs.begin(), nhs.end());
+      }
+    };
+
+    auto scale_pwhs = [&](std::vector<CGAL::Polygon_with_holes_2<EK>>& pwhs,
+                          const EK::FT& factor) {
+      for (auto& pwh : pwhs) {
+        CGAL::Polygon_2<EK> nb;
+        assert(pwh.outer_boundary().orientation() == CGAL::Sign::POSITIVE);
+        for (auto& point : pwh.outer_boundary()) {
+          nb.push_back(zero + (point - zero) * factor);
+        }
+        std::vector<CGAL::Polygon_2<EK>> nhs;
+        for (auto& hole : pwh.holes()) {
+          CGAL::Polygon_2<EK> nh;
+          for (auto& point : hole) {
+            assert(hole.orientation() == CGAL::Sign::NEGATIVE);
+            nb.push_back(zero + (point - zero) * factor);
+          }
+          nhs.push_back(std::move(nh));
+        }
+        pwh = CGAL::Polygon_with_holes_2<EK>(nb, nhs.begin(), nhs.end());
+      }
+    };
+
+    auto rotate_pwhs = [&](std::vector<CGAL::Polygon_with_holes_2<EK>>& pwhs,
+                           double turn) {
+      auto rotation = zturn2<EK>(turn);
+      for (auto& pwh : pwhs) {
+        CGAL::Polygon_2<EK> nb;
+        for (auto& point : pwh.outer_boundary()) {
+          nb.push_back(point.transform(rotation));
+        }
+        std::vector<CGAL::Polygon_2<EK>> nhs;
+        for (auto& hole : pwh.holes()) {
+          CGAL::Polygon_2<EK> nh;
+          for (auto& point : hole) {
+            nh.push_back(point.transform(rotation));
+          }
+          nhs.push_back(std::move(nh));
+        }
+        pwh = CGAL::Polygon_with_holes_2<EK>(nb, nhs.begin(), nhs.end());
+      }
+    };
+
+    auto setup_part = [&](Part& part) {
+      part.is_placed = false;
+
+      if (part.pwhs.size() < 1) {
+        return;
+      }
+      auto& pwh = part.pwhs[0];
+      if (pwh.outer_boundary().size() < 1) {
+        return;
+      }
+      part.area = 0;
+      part.pwhs_offset = EK::Point_2(0, 0) - pwh.outer_boundary()[0];
+      move_pwhs(part.pwhs, part.pwhs_offset);
+
+      // The npwhs is a reflection of the pwhs.
```

Comment:
These look like typos, I know they aren't but they look like it 😂

Reply:
Yeah, probably should have picked better names. :)